Google.Api.Generator/Generation/MetadataGenerator.cs

```diff
@@ -0,0 +1,70 @@
+﻿using Google.Gapic.Metadata;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json;
+
+namespace Google.Api.Generator.Generation
+{
+    internal static class MetadataGenerator
+    {
+        private const string GapicMetadataSchemaVersion = "1.0";
+        private const string GapicMetadataFileDescription = "This file maps proto services/RPCs to the corresponding library clients/methods";
+        private const string GapicMetadataLanguage = "csharp";
+
+        private const string TransportKeyGrpc = "grpc";
+
+        internal static string GenerateGapicMetadataJson(List<ServiceDetails> allServiceDetails)
+        {
+            
+            var gapicMetadata = new GapicMetadata
+            {
+                Schema = GapicMetadataSchemaVersion,
+                Comment = GapicMetadataFileDescription,
+                Language = GapicMetadataLanguage,
+                ProtoPackage = allServiceDetails.First().Package,
+                LibraryPackage = allServiceDetails.First().Namespace
+            };
+
+            foreach (var serviceDetails in allServiceDetails)
+            {
+                gapicMetadata.Services[serviceDetails.ServiceName] = ServiceForTransportMetadata(serviceDetails);
+            }
+
+            var gapicMetadataJson = JsonSerializer.Serialize(

```

Comment:
I would strongly recommend using the built-in JSON serialization in protobuf. It's entirely possible for there to be oddities that make proto messages not serialize/deserialize cleanly using Json.NET.

It's very reasonable to Json.NET for reformatting though.

Here's code that does that quite simply:

```csharp
string json = message.ToString();
JObject obj = JObject.Parse(json);
return obj.ToString(Formatting.Indented);
```

You *might* want to explicitly provide a settings object to prevent Json.NET from trying to parse dates.

Reply:
👍